### ISSUE TITLE
Support assembly code in cleo

### ISSUE THREAD

**Issue by XMDS** [sannybuilder/dev#275]
In fact, this may not be a good suggestion, but there are many cleo scripts doing patches and redirects (hooks), even on different architectures.  Usually we will write the byte of the assembly instruction through hex...end. This is not a good practice. First of all, it is very troublesome to write. Secondly, its code has poor readability. In addition, there are memory alignment problems on the arm architecture.  For the above reasons, it may be better to have sanny builder with a built-in assembly (disassembly) engine.

**Comment by MiranDMC** [comment]
Hm, why not use external asm compilator? You can then include its output file as hex block.

How would it work in Sanny? ASM compiler would be selected according to current edit mode?

**Comment by XMDS** [comment]
Implement disassembly using the Capstone Engine (which supports multiple programming languages):
https://github.com/capstone-engine/capstone.git

Implement assembly using the XB Engine (which supports multiple programming languages):
https://github.com/keystone-engine/keystone.git

@x87 I believe you can use these two mature and stable assembly engines, which support almost all mainstream CPU architectures and multiple programming languages.

**Comment by XMDS** [comment]
> Hm, why not use external asm compiler? You can then include its output file as hex block.
> 
> How would it work in Sanny? ASM compiler (platform) will be selected according to current edit mode?

In order to maintain code readability and convenience during development, and as you mentioned, to switch between different assembly engines based on Sanny's game mode, it is preferable to use an internal assembly compiler. This allows for seamless integration and efficient code management.

**Comment by MiranDMC** [comment]
> switch between different assembly engines based on Sanny's game mode, it is preferable to use an internal assembly compiler.

Assuming you create script that works on multiple platforms you will need few versions of assembly snippet for each target. Will it be possible to compile all of them (even if not used later anyway) without critical error stopping whole script compilation?

**Comment by XMDS** [comment]
> > switch between different assembly engines based on Sanny's game mode, it is preferable to use an internal assembly compiler.
> 
> Assuming you create script that works on multiple platforms you will need few versions of assembly snippet for each target. Will it be possible to compile all of them (even if not used later anyway) without critical error stopping whole script compilation?
> This would be perfectly avoidable with directives, which Sanny do not support.
> 
> 

这不是一个问题。只需要脚本编写者检测游戏平台和版本，然后为每种情况编写单独的分支即可。就像我们编写跨平台的CLEO脚本时一样，可以将每个平台和游戏分别处理。值得注意的是，例如Android和iOS的CLEO具有获取标签指针的单独命令，这与PC不同。无论如何，我们总是需要分别处理它们。

**Comment by MiranDMC** [comment]
Yes, but whole code is compiled, no matter what branching is used inside it. So all "ASM" blocks will be compiled then. Maybe ASM block should then contain preferred compiler in its definition.
